digid_eherkenning/models/base.py

```diff
@@ -166,7 +185,80 @@ class Meta:
     def __str__(self):
         return force_str(self._meta.verbose_name)
 
+    def populate_xml_fields(self, urls, xml):
+        """
+        Populates the idp_metadata_file and idp_service_entity_id fields based on the
+        fetched xml metadata
+        """
+        self.idp_service_entity_id = urls["entityId"]
+        content = ContentFile(xml.encode("utf-8"))
+        self.idp_metadata_file.save("metadata.xml", content, save=False)
+
+    def fetch_xml_metadata(self):
+        """
+        Retrieves the xml metadata from the provided url (metadata_file_source)
+
+        :return a string of the xml metadata.
+        """
+        return OneLogin_Saml2_IdPMetadataParser.get_metadata(self.metadata_file_source)
+
+    def parse_data_from_xml_source(self):
+        """
+        Parses the xml metadata
+
+        :return a tuple of a dictionary with the useful urls and the xml string itself.
+        """
+        urls = {}
+        xml = ""
+
+        try:
+            xml = self.fetch_xml_metadata()
+            parsed_idp_metadata = OneLogin_Saml2_IdPMetadataParser.parse(
+                xml,
+                required_sso_binding=OneLogin_Saml2_Constants.BINDING_HTTP_POST,
+                required_slo_binding=OneLogin_Saml2_Constants.BINDING_HTTP_POST,
+            )
+        except Exception:
+            raise ValidationError(
+                _(
+                    "An error has occured while processing the xml file. Make sure the file is valid"
+                )
+            )
+
+        if idp := parsed_idp_metadata.get("idp"):
+            # sometimes the xml file contains urn instead of a url as an entity ID
+            # use the provided url instead
+            urls = {
+                "entityId": (
+                    idp.get("entityId")
+                    if not idp.get("entityId").startswith("urn")
+                    else self.metadata_file_source
+                ),
+                "sso_url": idp.get("singleSignOnService", {}).get("url"),
+                "slo_url": idp.get("singleLogoutService", {}).get("url"),
+            }
+
+            if cache.get(self._meta.object_name):
+                cache.delete(self._meta.object_name)
+
+        return (urls, xml)
+
     def save(self, *args, **kwargs):
+        if self.pk and self.metadata_file_source:
+            if (
+                cache.get(self._meta.object_name, {}).get("entityId")
+                != self.metadata_file_source
+            ):
+                urls, xml = self.parse_data_from_xml_source()
+
+                if urls and xml:
```

Comment:
I think it'd make more sense if the processing raises an exception that no IDP was found rather than return empty containers to check. If there's no IDP in the metadata, that smells like a `ValidationError` to me because someone uploaded the wrong file or configured the wrong URL.

I would even go so far to do all this parsing/fetching/processing as part of the `def clean` so that it's part of the form validation in the admin.